### PR TITLE
docs(Building): add section for Arch

### PR DESCRIPTION
**Pull request by hasezoey**
This PR tries to add a section for Arch (and based on) for setting up a build environment.
The notes include what i had found setting it up on my system (Manjaro).

The install command may not include all required packages, because i had some previously installed (for other projects) so i dont quite know which relate to what / what needs to be installed extra than a base system

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9501)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Ping @yparitcher to review (because I seem to remember you runs Arch ;)).

**Comment by Frenzie** [comment]
We need ncurses to build? :thinking:

**Comment by hasezoey** [comment]
> We need ncurses to build? thinking

i dont see it in the build instructions, but when i tried to build it without `libtinfo.so.5` and `libncurses.so.5` existing, it errored out saying those files dont exist

**Comment by Frenzie** [comment]
What specifically errored? I believe that ncurses is an optional dependency for CMake, but it should work without. Adding more packages to the README may not be the right solution.

File: doc/Building.md [hunk 1]
```diff
@@ -34,6 +34,24 @@ Install the prerequisites using DNF:
...
+
+Also some manual changes have to be done:
+
+Arch only provides ncurses 6 now, but koreader needs `libncurses.so.5` and `libtinfo.so.5` but `ncurses` by default does not link them, so it has to be done manually:
+
+```sh
+sudo ln -sf /lib/libncursesw.so.6 /lib/libncurses.so.5
+sudo ln -sf /lib/libncursesw.so.6 /lib/libtinfo.so.5
+```
+
```
Comment (yparitcher) [inline doc/Building.md:54]: I have an up to date arch install.
ncurses included (bash along with half of my packages rely on it in some way)
I have never needed to symlink anything so this section should not be required. (Maybe the android SDK needs it but that is not a koreader issue)

**Review by yparitcher — CHANGES_REQUESTED** [review]

File: doc/Building.md (hunk 1 above)
Comment (yparitcher) [reply doc/Building.md:54]: cmake luarocks & git all depend on curl -> ca-certificates -> bash -> ncurses, and bash is part of the Arch base install. so it seems to me this section is superfluous.

The generic linux prerequisites should be enough.

File: doc/Building.md [hunk 2]
```diff
@@ -34,6 +34,24 @@ Install the prerequisites using DNF:
 sudo dnf install libstdc++-static SDL SDL-devel patch wget unzip git cmake luarocks autoconf nasm ragel gcc
```
Comment (yparitcher) [inline doc/Building.md:34]: not related but do we still need `libstdc++-static` as a dependency?

**Comment by hasezoey** [comment]
for testing i have removed `libtinfo.so.5` and `libncurses.so.5` (which are manual symlinks) again and run `kodev build android` and the error happens again

cutout from the log:
```txt
> Build command failed.
  Error while executing process /mnt/projects/android/koreader/base/toolchain/android-ndk-r15c/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/mnt/projects/android/koreader/platform/android/luajit-launcher/jni/Android.mk NDK_APPLICATION_MK=/mnt/projects/android/koreader/platform/android/luajit-launcher/jni/Application.mk APP_ABI=armeabi-v7a NDK_ALL_ABIS=armeabi-v7a NDK_DEBUG=0 APP_PLATFORM=android-14 NDK_OUT=/mnt/projects/android/koreader/platform/android/luajit-launcher/app/build/intermediates/ndkBuild/armRocksRelease/obj NDK_LIBS_OUT=/mnt/projects/android/koreader/platform/android/luajit-launcher/app/build/intermediates/ndkBuild/armRocksRelease/lib 7z luajit-launcher}
  make[3]: Entering directory `/mnt/projects/android/koreader/platform/android/luajit-launcher/app'
  make[3]: Nothing to be done for `7z'.
  [armeabi-v7a] Compile thumb  : luajit-launcher <= main.c
  make[3]: Leaving directory `/mnt/projects/android/koreader/platform/android/luajit-launcher/app'
  
  /mnt/projects/android/koreader/base/toolchain/android-ndk-r15c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory
  make[3]: *** [/mnt/projects/android/koreader/platform/android/luajit-launcher/app/build/intermediates/ndkBuild/armRocksRelease/obj/local/armeabi-v7a/objs/luajit-launcher/main.o] Error 127
```

[full log](https://github.com/koreader/koreader/files/9542052/log.txt)

**Comment by Frenzie** [comment]
> android

Well, there's your problem. ;-)

I don't know what's the correct method to set up an Android TC on Arch but that means that if necessary, it should go in this section instead.
https://github.com/koreader/koreader/blob/6f5c229c90fac58ce3b3c485dd9f16de518d951b/doc/Building_targets.md#additional-packages

File: doc/Building.md (hunk 1 above)
Comment (Frenzie) [reply doc/Building.md]: Why would Bash depend on ncurses? O.o

Bash itself may be an unwritten dependency though.

**Comment by hasezoey** [comment]
with 1858840fda0f09f46365cf5c519d0da08e63fe48 i have moved it to `Building_targets` and changed the wording

File: doc/Building_targets.md [hunk 3]
```diff
@@ -62,7 +62,14 @@ Building for Android requires `openjdk-8-jdk` and `p7zip-full`.
...
 ```
-sudo apt-get install openjdk-8-jdk p7zip-full
+sudo apt-get install openjdk-8-jdk p7zip-full ncurses
+```
+
+Note: In case a more up-to-date package manager is used (like Arch), then it may be necessary to link the following files because `ncurses` by default does not provide those anymore but are necessary for `android-ndk-r15c`:
+
+```sh
+sudo ln -sf ./libncursesw.so.6.3 ./libncurses.so.5
+sudo ln -sf ./libncursesw.so.6.3 ./libtinfo.so.5
```
Comment (Frenzie) [inline doc/Building_targets.md:72]: I don't know Arch very well but there seems to be a proper package instead?

https://aur.archlinux.org/packages/ncurses5-compat-libs

Comment (hasezoey) [reply doc/Building_targets.md:72]: there are 2 problems with that package (i had actually tried it):
1. it is a AUR so its not official
2. this package fails because its signature is not correct and seems to not have got and update to fix this

**Review by Frenzie — APPROVED** [review]
The proposed solution looks a little iffy to me but as far as docs structure goes lgtm.

File: doc/Building_targets.md (hunk 3 above)
Comment (NiLuJe) [reply doc/Building_targets.md:72]: It's... more complicated than that.

There *is* an option to keep building ncurses 6 w/ the ncurses 5 ABI, and the very visible result of that flags, besides the actual internal changes, is that it's built w/ the SOVER set to 5 ;).

TL;DR: If you don't have a `libncursesw.so.5`, your ncurses build does NOT, in fact, support the ncurses 5 ABI (and stuff *will* horribly crash at runtime).

**Comment by NiLuJe** [comment]
So, yeah, hard veto on the ncurses stuff until someone can find the Arch way of getting ncurses ABI5 binaries ;).

**Comment by NiLuJe** [comment]
FWIW, on Gentoo, we have an ncurses-compat ebuild (that currently builds ncurses 6.2), and on my system, its sole user is indeed...

```
 * These packages depend on ncurses-compat:
dev-util/android-sdk-update-manager-24.4.1-r1 (sys-libs/ncurses-compat:5[abi_x86_32(-)])
```

;o).

File: doc/Building_targets.md (hunk 3 above)
Comment (hasezoey) [reply doc/Building_targets.md:72]: > TL;DR: If you don't have a libncursesw.so.5, your ncurses build does NOT, in fact, support the ncurses 5 ABI (and stuff will horribly crash at runtime).

but i dont have it(only symlinked to 6.3) and it compiles and runs fine?

Comment (NiLuJe) [reply doc/Building_targets.md:72]: That's the magic of ABI mismatches: some stuff will work, some stuff will subtly break, some stuff will subtly do things differently, and some stuff will horribly crash ;).

Comment (Frenzie) [reply doc/Building_targets.md:72]: @NiLuJe I don't think you're interpreting it correctly, hence why I relented. It's just for clang to display compilation info in funky ways.

Comment (Frenzie) [reply doc/Building_targets.md:72]: I.e., ideally they shoudln't have built Clang with `-DLLVM_ENABLE_TERMINFO=on` or included the ncurses library in the NDK.

Comment (Frenzie) [reply doc/Building_targets.md:72]: Unless you think that it'll do things subtly differently *other than terminal output*?

Comment (NiLuJe) [reply doc/Building_targets.md:72]: I can't exclude the fact that it might horribly crash on *some* use-cases when it tries to poke at the wrong data structure or function in libtinfo, yeah :/.

Comment (NiLuJe) [reply doc/Building_targets.md:72]: Apparently, the usage seems trivial enough that it *might* be safe, but this is a *terrible* recommendation that may have system-wide side-effects ;).

Comment (NiLuJe) [reply doc/Building_targets.md:72]: If the symlink is the sanest solution Arch users can find, I'd isolate it to the toolchain itself instead of the system ;).

**Comment by yparitcher** [comment]
As per [ArchWiki](https://wiki.archlinux.org/title/android) Android page. The sdk / ndk packages are all in the AUR (Build it yourself) and most of them depend on the https://aur.archlinux.org/packages/ncurses5-compat-libs package @Frenzie linked too before. So that seems to be the way to do android

Like @NiLuJe, Hard veto

**Comment by hasezoey** [comment]
i agree that it is not a good solution, but it was necessary for me to get it to run

like i already had said earlier, https://aur.archlinux.org/packages/ncurses5-compat-libs is currently not working because of signature failures (and so i will not use it)

> If the symlink is the sanest solution Arch users can find, I'd isolate it to the toolchain itself instead of the system ;).

i also agree to that, though i dont quite know how to do this properly, because i also never really worked with c / c++ and dont like to mess with LD_PRELOAD or similar

---

i will close this pr, but it will exist for anyone also running into this issue as a guide (plus workaround)

File: doc/Building_targets.md (hunk 3 above)
Comment (Frenzie) [reply doc/Building_targets.md:72]: Yes!
